lib/pure/hashes.nim

```diff
@@ -56,6 +56,26 @@ type
 const
   IntSize = sizeof(int)
 
+proc preferStringHash*(T: typedesc): bool =
+  ## whether hashing is more efficient using `hash($x)`
+  # when string hashing is more efficient, see #11764
+  # exported so user defined hash can use this too
+  T.sizeof >= 4
+
+proc hash*(x: string): Hash
+
+proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} =
+  ## Efficient hashing of numbers, ordinals (eg enum), char.
+  # fix #11764
+  when preferStringHash(T):
+    when T is SomeFloat:
+      # needed otherwise finite precision can cause hash duplicate
+      let x = cast[BiggestInt](x)
+    hash($x)
```

Comment:
Sorry, I am not a fan of allocations. This is a string allocation. Why?

Reply:
because of https://github.com/nim-lang/Nim/pull/11767#issuecomment-512611465 ; but I found a different way to achieve same result (using jenkins hash instead) without allocation